lib/brakeman/warning.rb

```diff
@@ -53,6 +53,7 @@ def initialize options = {}
     OPTIONS.each do |key, var|
       self.instance_variable_set(var, options[key])
     end
+    @warning_type = @warning_type.dup
```

Comment:
Most of these look okay, but I'm a bit surprised by this one. I don't expect `warning_type` to get modified. Any clues to where that might be happening?

Reply:
It fails two tests:

```
  1) Error:
TestReportGeneration#test_tabs_sanity:
FrozenError: can't modify frozen String: "Cross-Site Scripting", created at /Users/byroot/src/github.com/casperisfine/brakeman/lib/brakeman/checks/check_content_tag.rb:199
    lib/brakeman/report/report_tabs.rb:12:in `gsub!'
    lib/brakeman/report/report_tabs.rb:12:in `block (2 levels) in generate_report'
    lib/brakeman/report/report_tabs.rb:10:in `map'
    lib/brakeman/report/report_tabs.rb:10:in `block in generate_report'
    lib/brakeman/report/report_tabs.rb:8:in `map'
    lib/brakeman/report/report_tabs.rb:8:in `generate_report'
    lib/brakeman/report.rb:107:in `generate'
    lib/brakeman/report.rb:58:in `format'
    lib/brakeman/report.rb:63:in `method_missing'
    test/tests/report_generation.rb:67:in `test_tabs_sanity'

  2) Error:
TestTabsOutput#test_reported_warnings:
FrozenError: can't modify frozen String: "Cross-Site Request Forgery", created at /Users/byroot/src/github.com/casperisfine/brakeman/lib/brakeman/checks/check_csrf_token_forgery_cve.rb:19
    lib/brakeman/report/report_tabs.rb:12:in `gsub!'
    lib/brakeman/report/report_tabs.rb:12:in `block (2 levels) in generate_report'
    lib/brakeman/report/report_tabs.rb:10:in `map'
    lib/brakeman/report/report_tabs.rb:10:in `block in generate_report'
    lib/brakeman/report/report_tabs.rb:8:in `map'
    lib/brakeman/report/report_tabs.rb:8:in `generate_report'
    lib/brakeman/report.rb:107:in `generate'
    lib/brakeman/report.rb:58:in `format'
    lib/brakeman/report.rb:63:in `method_missing'
    test/tests/tabs_output.rb:9:in `setup'
```

```ruby
class Brakeman::Report::Tabs < Brakeman::Report::Table
  def generate_report
    [[:generic_warnings, "General"], [:controller_warnings, "Controller"],
      [:model_warnings, "Model"], [:template_warnings, "Template"]].map do |meth, category|

      self.send(meth).map do |w|
        line = w.line || 0
        w.warning_type.gsub!(/[^\w\s]/, ' ')
        "#{(w.file.absolute)}\t#{line}\t#{w.warning_type}\t#{category}\t#{w.format_message}\t#{w.confidence_name}"
      end.join "\n"

    end.join "\n"

  end
end
```

It very explicitly mutate that string in place, hence why I chose to make it mutable. But to be honest it looks like a bug, but I don't know brakeman enough to decide.

If I delete that `gsub!` the test suite still pass.

Reply:
Let's delete the `gsub!`. Not sure what I thought it was doing... but it doesn't. Also I plan to deprecate+remove that report format.